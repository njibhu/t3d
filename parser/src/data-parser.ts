import { BaseType, DataType } from "./types";

interface Definition {
  definitions?: {
    [definition: string]: { [key: string]: DataType | string };
  };
  root: {
    [key: string]: DataType | string;
  };
}

interface ParseFunctionReturn {
  newPosition: number;
  data: any;
}

export class DataParser implements Definition {
  public readonly definitions: Definition["definitions"];
  public readonly root: Definition["root"];
  public DEBUG: boolean;

  constructor(definition: Definition) {
    Object.assign(this, definition);
    this.DEBUG = false;
  }

  public parse(dv: DataView, pos: number): ParseFunctionReturn {
    let position = pos;
    const parsedObject = {};
    for (const key in this.root) {
      const value = this.root[key];
      let parsedResult: ParseFunctionReturn;
      if (typeof value === "string") {
        if (this.DEBUG) console.log(key, position.toString(16), "(parseType)");
        parsedResult = this.parseType(dv, position, value);
      } else {
        const { baseType, subType, length } = value;
        if (this.DEBUG) console.log(key, position.toString(16), baseType, subType, length);
        parsedResult = this[baseType](dv, position, subType, length);
      }
      parsedObject[key] = parsedResult.data;
      position = parsedResult.newPosition;
    }

    return {
      newPosition: position,
      data: parsedObject,
    };
  }

  public parseType(dv: DataView, pos: number, typeDefinitionName: string): ParseFunctionReturn {
    const parsedObject = {};
    let position = pos;
    const definition = this.definitions[typeDefinitionName];
    for (const key in definition) {
      const value = definition[key];
      let parsedResult: ParseFunctionReturn;
      if (typeof value === "string") {
        if (this.DEBUG) console.log(">", key, position.toString(16), "(parseType)");
        parsedResult = this.parseType(dv, position, value);
      } else {
        const { baseType, subType, length } = value;
        if (this.DEBUG) console.log(">", key, position.toString(16), baseType, subType, length);
        parsedResult = this[baseType](dv, position, subType, length);
      }
      parsedObject[key] = parsedResult.data;

      position = parsedResult.newPosition;
    }

    return {
      newPosition: position,
      data: parsedObject,
    };
  }

  /**
   *      Parsers
   **/

  private Float32(dv: DataView, pos: number): ParseFunctionReturn {
    if (this.DEBUG) console.debug("Float32", dv.getFloat32(pos, true));
    return { newPosition: pos + 4, data: dv.getFloat32(pos, true) };
  }

  private Float64(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 8, data: dv.getFloat64(pos, true) };
  }

  private Uint8(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 1, data: dv.getUint8(pos) };
  }

  private Uint16(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 2, data: dv.getUint16(pos, true) };
  }

  private Uint32(dv: DataView, pos: number): ParseFunctionReturn {
    if (this.DEBUG) console.debug("Uint32", dv.getUint32(pos, true));
    return { newPosition: pos + 4, data: dv.getUint32(pos, true) };
  }

  private Uint64(dv: DataView, pos: number): ParseFunctionReturn {
    if (this.DEBUG) console.debug("Uint64", dv.getUint32(pos, true), dv.getUint32(pos + 4, true));
    return {
      newPosition: pos + 8,
      data: (BigInt(dv.getUint32(pos + 4, true)) << BigInt(32)) | BigInt(dv.getUint32(pos, true)),
    };
  }

  private CString(dv: DataView, pos: number, _subType: DataType | string, length: number): ParseFunctionReturn {
    const u8 = new Uint8Array(dv.buffer, pos);
    const end = length || u8.findIndex((v) => v === 0);

    if (this.DEBUG) console.debug("CString", String.fromCharCode.apply(null, new Uint8Array(u8.slice(0, end))));
    return {
      newPosition: pos + end,
      data: String.fromCharCode.apply(null, new Uint8Array(u8.slice(0, end))),
    };
  }

  private RefString(dv: DataView, pos: number): ParseFunctionReturn {
    let ptr = pos + dv.getUint32(pos, true);

    let data = "";
    let num;
    while (ptr + 1 < dv.byteLength && (num = dv.getUint8(ptr)) !== 0) {
      ptr += 1;
      data += String.fromCharCode(num);
    }

    return {
      newPosition: pos + 4,
      data,
    };
  }

  private FixedArray(dv: DataView, pos: number, type: DataType | string, length: number): ParseFunctionReturn {
    // Some types can be mapped directly from their buffer into the return type
    if (typeof type != "string" && this._getOptimisedArrayConstructor(type.baseType)) {
      return this.optimisedArray(dv, pos, type, length);
    }

    const data = [];
    let newPosition = pos;
    for (let itemIndex = 0; itemIndex < length; itemIndex++) {
      const parsedItem =
        typeof type === "string"
          ? this.parseType(dv, newPosition, type)
          : this[type.baseType](dv, newPosition, type.subType, type.length);
      data.push(parsedItem.data);
      newPosition = parsedItem.newPosition;
    }

    return {
      newPosition,
      data,
    };
  }

  private DynArray(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    let arrayLength = dv.getUint32(pos, true);
    let arrayOffset = dv.getUint32(pos + 4, true);
    if (this.DEBUG) console.debug("DynArray", arrayLength, arrayOffset);

    if (arrayOffset === 0) {
      return {
        newPosition: pos + 8,
        data: [],
      };
    }
    let arrayPtr = pos + 4 + arrayOffset;

    return {
      newPosition: pos + 8,
      data: this.FixedArray(dv, arrayPtr, type, arrayLength).data,
    };
  }

  private RefArray(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    let data = [];
    let arrayLength = dv.getUint32(pos, true);
    let arrayPtr = dv.getUint32(pos + 4, true) + pos + 4;
    if (this.DEBUG) console.debug("RefArray", "arrayPtr", (pos + 4).toString(16), arrayPtr.toString(16));
    if (arrayLength === 0) {
      return {
        data,
        newPosition: pos + 8,
      };
    }

    const pointer = dv.getUint32(pos + 4, true) + pos + 4;
    for (let index = 0; index < arrayLength; index++) {
      const offset = dv.getUint32(arrayPtr + 4 * index, true);
      if (offset !== 0) {
        if (this.DEBUG) console.debug("RefArray", "offset", offset.toString(16));
        let newPosition = pointer + index * 4 + offset;
        if (typeof type === "string") {
          data.push(this.parseType(dv, newPosition, type).data);
        } else {
          data.push(this[type.baseType](dv, newPosition, type.subType, type.length).data);
        }
      }
    }

    return {
      newPosition: pos + 8,
      data,
    };
  }

  private Pointer(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    const offset = dv.getUint32(pos, true);
    const parsedItem =
      typeof type === "string"
        ? this.parseType(dv, pos + offset, type)
        : this[type.baseType](dv, pos + offset, type.subType, type.length);

    return {
      newPosition: pos + 4,
      data: parsedItem.data,
    };
  }

  private RefString16(dv: DataView, pos: number): ParseFunctionReturn {
    let ptr = pos + dv.getUint32(pos, true);

    let data = "";
    let num;
    while (ptr + 2 < dv.byteLength && (num = dv.getUint16(ptr, true)) !== 0) {
      ptr += 2;
      data += String.fromCharCode(num);
    }

    return {
      newPosition: pos + 4,
      data,
    };
  }

  private Fileref(dv: DataView, pos: number): ParseFunctionReturn {
    return this.Filename(dv, pos);
  }

  private Filename(dv: DataView, pos: number): ParseFunctionReturn {
    // This implementation is based on the old Utils.getFileNameReader() function
    let newPosition = pos;
    //try {
    let ptr = newPosition + dv.getUint32(newPosition, true);
    newPosition += 4;

    const m_lowPart = dv.getUint16(ptr, true);
    ptr += 2;
    const m_highPart = dv.getUint16(ptr, true);
    ptr += 2;
    const _m_terminator = dv.getUint16(ptr, true);
    ptr += 2;

    /// Getting the file name...
    /// Both need to be >= than 256 (terminator is 0)
    const ret = 0xff00 * (m_highPart - 0x100) + (m_lowPart - 0x100) + 1;

    return {
      newPosition: pos + 4,
      data: ret > 0 ? ret : 0,
    };
    // } catch (e) {
    //   return {
    //     newPosition,
    //     data: -1,
    //   };
    // }
  }

  private Unknown(dv: DataView, pos: number): ParseFunctionReturn {
    throw new Error("Could not parse unknown data");
  }

  /**
   *      Parser utils & helpers
   **/

  private optimisedArray(dv: DataView, pos: number, type: DataType, length: number): ParseFunctionReturn {
    const OptimisedArray = this._getOptimisedArrayConstructor(type.baseType);
    if (this.DEBUG) console.debug("OptimizedArray", Array.from(new OptimisedArray(dv.buffer, pos, length)));
    return {
      newPosition: pos + length * OptimisedArray.BYTES_PER_ELEMENT,
      data: Array.from(new OptimisedArray(dv.buffer, pos, length)),
    };
  }

  private _getOptimisedArrayConstructor(baseType: BaseType) {
    switch (baseType) {
      case BaseType.Float32:
        return Float32Array;

      case BaseType.Float64:
        return Float64Array;

      case BaseType.Uint8:
        return Uint8Array;

      case BaseType.Uint16:
        return Uint16Array;

      // case BaseType.Uint32:
      //   return Uint32Array;
    }
  }
}
