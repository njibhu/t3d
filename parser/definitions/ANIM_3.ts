import { Uint64, Fileref, Float32, Uint32, DynArray } from "../src/types";

export const V0 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV0",
  version: 0,
  definitions: {
    PackEmoteAnimationV0: {
      token: Uint64,
      timings: DynArray("PackEmoteTimingV0")
    },
    PackEmoteTimingV0: {
      modelFileId: Fileref(),
      blendInTime: Float32,
      blendOutTime: Float32,
      duration: Uint32
    }
  },
  root: {
    animations: DynArray("PackEmoteAnimationV0")
  }
};

export const V1 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV1",
  version: 1,
  definitions: {
    PackEmoteAnimationV1: {
      token: Uint64,
      timings: DynArray("PackEmoteTimingV1")
    },
    PackEmoteTimingV1: {
      modelFileId: Fileref(),
      blendInTime: Float32,
      blendOutTime: Float32,
      duration: Uint32,
      loopDuration: Uint32
    }
  },
  root: {
    animations: DynArray("PackEmoteAnimationV1")
  }
};

export const V2 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV2",
  version: 2,
  definitions: {
    PackEmoteAnimationV2: {
      Token: Uint64,
      Timing: DynArray("PackEmoteTimingV2")
    },
    PackEmoteTimingV2: {
      ModelFile: Fileref(),
      BlendIn: Float32,
      BlendOut: Float32,
      IntroDuration: Uint32,
      LoopDuration: Uint32,
      OutroDuration: Uint32
    }
  },
  root: {
    Animation: DynArray("PackEmoteAnimationV2")
  }
};

export const V3 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV3",
  version: 3,
  definitions: {
    PackEmoteAnimationV3: {
      Token: Uint64,
      Timing: DynArray("PackEmoteTimingV3")
    },
    PackEmoteTimingV3: {
      ModelFile: Fileref(),
      BlendIn: Float32,
      BlendOut: Float32,
      IntroDuration: Uint32,
      LoopDuration: Uint32,
      OutroDuration: Uint32,
      StartOffset: Uint32
    }
  },
  root: {
    Animation: DynArray("PackEmoteAnimationV3")
  }
};

export const latest = V3;
export const definitionArray = [V0, V1, V2, V3];