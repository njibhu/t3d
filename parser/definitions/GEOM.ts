import { Uint64, Uint16, FixedArray, Float32, DynArray, Uint32, CString, Uint8, Pointer, RefArray } from "../src/types";

export const V0 = {
  chunkName: "GEOM",
  name: "ModelFileGeometryV0",
  version: 0,
  definitions: {
    ModelMeshDataV65: {
      visBone: Uint64,
      morphTargets: DynArray("ModelMeshMorphTargetV65"),
      flags: Uint32,
      seamVertIndices: DynArray(Uint32),
      meshName: Uint64,
      minBound: FixedArray(Float32, 3),
      maxBound: FixedArray(Float32, 3),
      bounds: DynArray("GrBoundData"),
      materialIndex: Uint32,
      materialName: CString(),
      boneNames: DynArray(CString()),
      geometry: Pointer("ModelMeshGeometryV0")
    },
    ModelMeshMorphTargetV65: {
      positions: DynArray("ModelMeshMorphVertV65"),
      normals: DynArray("ModelMeshMorphVertV65"),
      mesh: Uint64
    },
    ModelMeshMorphVertV65: {
      index: Uint16,
      vector: FixedArray(Float32, 3)
    },
    GrBoundData: {
      center: FixedArray(Float32, 3),
      boxExtent: FixedArray(Float32, 3),
      sphereRadius: Float32
    },
    ModelMeshGeometryV0: {
      verts: "ModelMeshVertexDataV0",
      indices: "ModelMeshIndexDataV0",
      lods: DynArray("ModelMeshIndexDataV0")
    },
    ModelMeshVertexDataV0: {
      vertexCount: Uint32,
      mesh: "PackVertexType"
    },
    PackVertexType: {
      fvf: Uint32,
      vertices: DynArray(Uint8)
    },
    ModelMeshIndexDataV0: {
      indices: DynArray(Uint16)
    }
  },
  root: {
    meshes: RefArray("ModelMeshDataV65")
  }
};

export const V1 = {
  chunkName: "GEOM",
  name: "ModelFileGeometryV1",
  version: 1,
  definitions: {
    ModelMeshDataV66: {
      visBone: Uint64,
      morphTargets: DynArray("ModelMeshMorphTargetV66"),
      flags: Uint32,
      seamVertIndices: DynArray(Uint32),
      meshName: Uint64,
      minBound: FixedArray(Float32, 3),
      maxBound: FixedArray(Float32, 3),
      bounds: DynArray("GrBoundData"),
      materialIndex: Uint32,
      materialName: CString(),
      boneBindings: DynArray(Uint64),
      geometry: Pointer("ModelMeshGeometryV1")
    },
    ModelMeshMorphTargetV66: {
      positions: DynArray("ModelMeshMorphVertV66"),
      normals: DynArray("ModelMeshMorphVertV66"),
      mesh: Uint64
    },
    ModelMeshMorphVertV66: {
      index: Uint16,
      vector: FixedArray(Float32, 3)
    },
    GrBoundData: {
      center: FixedArray(Float32, 3),
      boxExtent: FixedArray(Float32, 3),
      sphereRadius: Float32
    },
    ModelMeshGeometryV1: {
      verts: "ModelMeshVertexDataV1",
      indices: "ModelMeshIndexDataV1",
      lods: DynArray("ModelMeshIndexDataV1"),
      transforms: DynArray(Uint32)
    },
    ModelMeshVertexDataV1: {
      vertexCount: Uint32,
      mesh: "PackVertexType"
    },
    PackVertexType: {
      fvf: Uint32,
      vertices: DynArray(Uint8)
    },
    ModelMeshIndexDataV1: {
      indices: DynArray(Uint16)
    }
  },
  root: {
    meshes: RefArray("ModelMeshDataV66")
  }
};

export const latest = V1;
export const definitionArray = [V0, V1];