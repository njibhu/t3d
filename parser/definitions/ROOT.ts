import { Uint64, Float32, DynArray, FixedArray } from "../src/types";

export const V0 = {
  chunkName: "ROOT",
  name: "ModelFileRootMotionV0",
  version: 0,
  definitions: {
    ModelRootMotionV0: {
      sequence: Uint64,
      keys: DynArray(Float32),
      values: DynArray(FixedArray(Float32, 3))
    }
  },
  root: {
    rootMotions: DynArray("ModelRootMotionV0")
  }
};

export const V1 = {
  chunkName: "ROOT",
  name: "ModelFileRootMotionV1",
  version: 1,
  definitions: {
    ModelRootMotionV1: {
      sequence: Uint64,
      keys: DynArray(Float32),
      posValues: DynArray(FixedArray(Float32, 3)),
      quatValues: DynArray(FixedArray(Float32, 4))
    }
  },
  root: {
    rootMotions: DynArray("ModelRootMotionV1")
  }
};

export const latest = V1;
export const definitionArray = [V0, V1];