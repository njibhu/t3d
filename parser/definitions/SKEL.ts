import { CString, Uint32, FixedArray, Float32, Uint8, Pointer, DynArray, Uint64, Uint16, Filename } from "../src/types";

export const V0 = {
  chunkName: "SKEL",
  name: "ModelFileSkeletonV0",
  version: 0,
  definitions: {
    ModelSkeletonDataV62: {
      grannyModel: Pointer("ModelGrannyModelV0"),
      boneConstraints: DynArray("ModelBoneConstraintV62"),
      boneFlags: DynArray(Uint32),
      boneSymmetries: DynArray("ModelBoneSymmetryV62"),
      emitterBones: DynArray(Uint32),
      trackMasks: DynArray("ModelTrackMaskV62")
    },
    ModelGrannyModelV0: {
      Name: CString(),
      Skeleton: Pointer("ModelGrannySkeletonV0"),
      InitialPlacement: "ModelTransformData",
      MeshBindings: DynArray("ModelMeshBindingData"),
      ExtendedData: Pointer(Uint8)
    },
    ModelGrannySkeletonV0: {
      Name: CString(),
      Bones: DynArray("ModelBoneData"),
      LODType: Uint32,
      ExtendedData: Pointer(Uint8)
    },
    ModelBoneData: {
      Name: CString(),
      ParentIndex: Uint32,
      LocalTransform: "ModelTransformData",
      InverseWorld4x4: FixedArray(FixedArray(Float32, 4), 4),
      LODError: Float32,
      ExtendedData: Pointer(Uint8)
    },
    ModelTransformData: {
      Flags: Uint32,
      Position: FixedArray(Float32, 3),
      Orientation: FixedArray(Float32, 4),
      ScaleShear: FixedArray(FixedArray(Float32, 3), 3)
    },
    ModelMeshBindingData: {
      Mesh: Pointer(Uint8)
    },
    ModelBoneConstraintV62: {
      token: Uint64,
      flags: Uint16,
      twistOffset: Float32,
      animBlend: Float32,
      drag: Float32,
      ellipseRatio: Float32,
      gravity: Float32,
      collisionRadius: Float32,
      wind: Float32,
      angle: FixedArray(Float32, 2),
      angleStrength: Float32,
      angleType: Uint8,
      distanceInner: FixedArray(Float32, 2),
      distanceInnerStrength: Float32,
      distanceInnerType: Uint8,
      links: DynArray("ModelBoneConstraintLinkV62"),
      distanceOuter: FixedArray(Float32, 2),
      distanceOuterStrength: Float32,
      distanceOuterType: Uint8,
      twist: FixedArray(Float32, 2),
      twistStrength: Float32,
      twistType: Uint8
    },
    ModelBoneConstraintLinkV62: {
      angle: Float32,
      azimuth: Float32,
      distance: FixedArray(Float32, 2),
      token: Uint64
    },
    ModelBoneSymmetryV62: {
      boneLeft: Uint64,
      boneRight: Uint64
    },
    ModelTrackMaskV62: {
      data: "PackGrannyTrackMaskType",
      token: Uint64
    },
    PackGrannyTrackMaskType: {
      trackMask: DynArray(Uint8)
    }
  },
  root: {
    skeletonData: Pointer("ModelSkeletonDataV62"),
    fileReference: Filename()
  }
};

export const V1 = {
  chunkName: "SKEL",
  name: "ModelFileSkeletonV1",
  version: 1,
  definitions: {
    ModelSkeletonDataV63: {
      grannyModel: Pointer("ModelGrannyModelV1"),
      boneConstraints: DynArray("ModelBoneConstraintV63"),
      boneFlags: DynArray(Uint32),
      mirrorSpec: "PackGrannyMirrorSpecType",
      emitterBones: DynArray(Uint32),
      trackMasks: DynArray("ModelTrackMaskV63")
    },
    ModelGrannyModelV1: {
      Name: CString(),
      Skeleton: Pointer("ModelGrannySkeletonV1"),
      InitialPlacement: "ModelTransformData",
      MeshBindings: DynArray("ModelMeshBindingData"),
      ExtendedData: Pointer(Uint8)
    },
    ModelGrannySkeletonV1: {
      Name: CString(),
      Bones: DynArray("ModelBoneData"),
      LODType: Uint32,
      ExtendedData: Pointer(Uint8)
    },
    ModelBoneData: {
      Name: CString(),
      ParentIndex: Uint32,
      LocalTransform: "ModelTransformData",
      InverseWorld4x4: FixedArray(FixedArray(Float32, 4), 4),
      LODError: Float32,
      ExtendedData: Pointer(Uint8)
    },
    ModelTransformData: {
      Flags: Uint32,
      Position: FixedArray(Float32, 3),
      Orientation: FixedArray(Float32, 4),
      ScaleShear: FixedArray(FixedArray(Float32, 3), 3)
    },
    ModelMeshBindingData: {
      Mesh: Pointer(Uint8)
    },
    ModelBoneConstraintV63: {
      token: Uint64,
      flags: Uint16,
      twistOffset: Float32,
      animBlend: Float32,
      drag: Float32,
      ellipseRatio: Float32,
      gravity: Float32,
      collisionRadius: Float32,
      wind: Float32,
      angle: FixedArray(Float32, 2),
      angleStrength: Float32,
      angleType: Uint8,
      distanceInner: FixedArray(Float32, 2),
      distanceInnerStrength: Float32,
      distanceInnerType: Uint8,
      links: DynArray("ModelBoneConstraintLinkV63"),
      distanceOuter: FixedArray(Float32, 2),
      distanceOuterStrength: Float32,
      distanceOuterType: Uint8,
      twist: FixedArray(Float32, 2),
      twistStrength: Float32,
      twistType: Uint8
    },
    ModelBoneConstraintLinkV63: {
      angle: Float32,
      azimuth: Float32,
      distance: FixedArray(Float32, 2),
      token: Uint64
    },
    PackGrannyMirrorSpecType: {
      mirrorSpec: DynArray(Uint8)
    },
    ModelTrackMaskV63: {
      data: "PackGrannyTrackMaskType",
      token: Uint64
    },
    PackGrannyTrackMaskType: {
      trackMask: DynArray(Uint8)
    },
    ModelSkeletonOverridesV1: {
      boneConstraints: DynArray("ModelBoneConstraintV63")
    }
  },
  root: {
    skeletonData: Pointer("ModelSkeletonDataV63"),
    fileReference: Filename(),
    overrides: Pointer("ModelSkeletonOverridesV1")
  }
};

export const latest = V1;
export const definitionArray = [V0, V1];