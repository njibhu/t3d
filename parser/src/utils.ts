import { DataParser } from "./data-parser";
import { ChunkDefinitions } from "./struct-definitions";
import type { ChunkHead, FileHead } from "./interfaces";

const __IS_NODE_RUNTIME = typeof process !== "undefined" && process.versions != null && process.versions.node != null;

export function sliceFile(file: File, offset: number, length: number): Promise<ArrayBuffer> {
  if (__IS_NODE_RUNTIME) {
    // @ts-expect-error - Node specific code
    return import("node:fs").then((fs) => {
      const fd = fs.openSync(file);
      const buffer = Buffer.alloc(length);
      fs.readSync(fd, buffer, 0, length, offset);
      fs.closeSync(fd);
      return buffer.buffer;
    });
  } else {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();

      reader.onload = () => {
        resolve(reader.result as ArrayBuffer);
      };

      reader.onerror = reject;

      reader.readAsArrayBuffer(file.slice(Number(offset), Number(offset) + Number(length)));
    });
  }
}

export function parseFile(dataView: DataView): { newPosition: number; data: FileHead } {
  const fileHeaderParser = new DataParser({ root: ChunkDefinitions.FILE_HEAD });
  const result = fileHeaderParser.parse(dataView, 0);
  result.data.type = result.data.type.replace("\u0000", "");
  return result;
}

export function parseChunkHead(dataView: DataView, pos: number): { newPosition: number; data: ChunkHead } {
  const chunkHeadParser = new DataParser({ root: ChunkDefinitions.CHUNK_HEAD });
  const result = chunkHeadParser.parse(dataView, pos);
  result.data.type = result.data.type.replace("\u0000", "");
  return result;
}

export function parseAllChunks(
  dataView: DataView,
  firstChunkPosition: number
): { chunkPosition: number; chunkHeader: ChunkHead }[] {
  const chunks = [];
  let pos = firstChunkPosition;
  while (pos < dataView.byteLength) {
    const chunk = parseChunkHead(dataView, pos);
    chunks.push({ chunkPosition: pos, chunkHeader: chunk.data });
    pos = chunk.newPosition - 8 + chunk.data.chunkDataSize;
  }
  return chunks;
}
