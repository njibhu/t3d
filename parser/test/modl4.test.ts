import * as fs from "fs";

import { DataParser } from "../src/data-parser";
import { parseFile, parseAllChunks } from "../src/utils";

import {  toArrayBuffer } from "./test-helper";

import * as MODL from "../definitions/MODL";
import * as GEOM from "../definitions/GEOM";
import * as ROOT from "../definitions/ROOT";
import * as SKEL from "../definitions/SKEL";

const chunkBuffer = fs.readFileSync("./test/content/modl4.bin", null);
const dv = new DataView(toArrayBuffer(chunkBuffer));
const fileHead = parseFile(dv);
const allChunks = parseAllChunks(dv, fileHead.newPosition);
const is64Bit = fileHead.data.flags == 5;

describe("modl4", () => {
  test("matches for MODL", function () {
    const modlChunk = allChunks.find((c) => c.chunkHeader.type === "MODL");
    const def = MODL.definitions[`V${modlChunk!.chunkHeader.chunkVersion}` as keyof typeof MODL["definitions"]];
    const parser = new DataParser(def, is64Bit);
    //parser.DEBUG = true;
    const test = parser.parse(dv, modlChunk!.chunkPosition + modlChunk!.chunkHeader.chunkHeaderSize);
    expect(test.data).toMatchSnapshot("modl4-modl")
  });

  // GEOM
  test("matches for GEOM", function () {
    const geomChunk = allChunks.find((c) => c.chunkHeader.type === "GEOM");
    const def = GEOM.definitions[`V${geomChunk!.chunkHeader.chunkVersion}` as keyof typeof GEOM["definitions"]];
    const parser = new DataParser(def, is64Bit);
    //parser.DEBUG = true;
    const test = parser.parse(dv, geomChunk!.chunkPosition + geomChunk!.chunkHeader.chunkHeaderSize);

    expect(test.data).toMatchSnapshot("modl4-geom")
  });

  //ROOT
  test("matches for ROOT", function () {
    const rootChunk = allChunks.find((c) => c.chunkHeader.type === "ROOT");
    const def = ROOT.definitions[`V${rootChunk!.chunkHeader.chunkVersion}` as keyof typeof ROOT["definitions"]];
    const parser = new DataParser(def, is64Bit);
    //parser.DEBUG = true;
    const test = parser.parse(dv, rootChunk!.chunkPosition + rootChunk!.chunkHeader.chunkHeaderSize);
    expect(test.data).toMatchSnapshot("modl4-root")
  });

  //SKEL
  test("matches for SKEL", function () {
    const skelChunk = allChunks.find((c) => c.chunkHeader.type === "SKEL");
    const def = SKEL.definitions[`V${skelChunk!.chunkHeader.chunkVersion}` as keyof typeof SKEL["definitions"]];
    const test = new DataParser(def, is64Bit).parse(dv, skelChunk!.chunkPosition + skelChunk!.chunkHeader.chunkHeaderSize);
    expect(test.data).toMatchSnapshot("modl2-skel")
  });
});
