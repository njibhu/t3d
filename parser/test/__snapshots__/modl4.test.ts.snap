// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modl4 matches for GEOM: modl4-geom 1`] = `
{
  "meshes": [],
}
`;

exports[`modl4 matches for MODL: modl4-modl 1`] = `
{
  "actionOffsetNames": [],
  "actionOffsets": [],
  "boneOffsetData": [],
  "boundingSphere": {
    "center": Float32Array [
      218.6875,
      -0.1292724609375,
      0,
    ],
    "radius": 219.69456481933594,
  },
  "clothData": [],
  "cloudData": {},
  "lightData": {},
  "lightningData": {
    "bolts": [
      {
        "bone": 1152921504607515106n,
        "colorBegin": Uint32Array [
          4286544752,
          4285690482,
          4286544752,
        ],
        "colorEnd": Uint32Array [
          4285690482,
          4286544752,
          4285690482,
        ],
        "colorFalloff": Float32Array [
          0,
          1,
        ],
        "colorPeriod": 0.16666699945926666,
        "flags": 0,
        "flipbook": {},
        "fps": 0,
        "frequency": 3,
        "groupMax": 1,
        "groupMin": 1,
        "lifetime": Float32Array [
          1.5,
          1.5,
        ],
        "nodeIndices": Uint16Array [
          1,
          0,
        ],
        "noise": 0,
        "numSegments": 10,
        "opacity": Float32Array [
          1,
          1,
        ],
        "opacityPreset": 2,
        "probability": 100,
        "texOffset": 1,
        "texRange": Float32Array [
          0,
          1,
        ],
        "texScale": 300,
        "texSpeed": -0.10000000149011612,
        "thicknessPreset": 0,
        "thicknessRange": Float32Array [
          40,
          45,
        ],
        "type": 0,
        "variance": 0,
        "variancePreset": 0,
      },
    ],
    "nodes": [
      {
        "bone": 2305843185826157958n,
        "childrenIndices": [],
        "flags": 0,
        "probability": 100,
        "radius": Float32Array [
          0,
          1,
        ],
        "shape": 0,
        "updatePos": 1,
      },
      {
        "bone": 1152921681219310982n,
        "childrenIndices": Uint16Array [
          0,
        ],
        "flags": 2,
        "probability": 100,
        "radius": Float32Array [
          0,
          1,
        ],
        "shape": 0,
        "updatePos": 1,
      },
    ],
    "systems": [
      {
        "boltIndices": Uint16Array [
          0,
        ],
        "bone": 1152921505048973107n,
        "flags": 0,
        "fvf": 1817,
        "materialIndex": 0,
      },
    ],
  },
  "lodOverride": Float32Array [
    3500,
    7500,
  ],
  "obstacles": [],
  "permutations": [
    {
      "materials": [
        {
          "constants": [
            {
              "constantFlags": 2,
              "name": 1061534053,
              "value": Float32Array [
                0.10000000149011612,
                0,
                1.401298464324817e-45,
                0,
              ],
            },
            {
              "constantFlags": 2,
              "name": 1416822620,
              "value": Float32Array [
                1.0019999742507935,
                0,
                1.401298464324817e-45,
                0,
              ],
            },
            {
              "constantFlags": 2,
              "name": 805479253,
              "value": Float32Array [
                0.47916701436042786,
                0,
                1.401298464324817e-45,
                0,
              ],
            },
          ],
          "filename": 543181,
          "matConstLinks": [
            {
              "constantToken": 1061534053,
              "linkToken": 0n,
            },
            {
              "constantToken": 1416822620,
              "linkToken": 0n,
            },
            {
              "constantToken": 805479253,
              "linkToken": 0n,
            },
          ],
          "materialFlags": 520,
          "materialId": 6,
          "sortOrder": 0,
          "texCoordCount": 2,
          "texTransforms": [
            {
              "columns": 1,
              "count": 1,
              "flags": 0,
              "fps": 0,
              "rotate": 0,
              "rotate_": 0,
              "rows": 1,
              "scale": Float32Array [
                0,
                0,
              ],
              "scaleFreq": Float32Array [
                0,
                0,
              ],
              "scroll": Float32Array [
                -0.10000000149011612,
                0,
              ],
              "scrollFreq": Float32Array [
                0,
                0,
              ],
              "uvIndex": 1,
            },
          ],
          "textures": [
            {
              "blitId": 0n,
              "filename": 1015139,
              "textureFlags": 0,
              "token": 27219515885689124n,
              "uvAnimId": 0,
              "uvPSInputIndex": 0,
            },
            {
              "blitId": 0n,
              "filename": 1015140,
              "textureFlags": 0,
              "token": 27219512631380144n,
              "uvAnimId": 0,
              "uvPSInputIndex": 1,
            },
          ],
          "token": 774n,
          "uvTransLinks": [],
        },
      ],
      "token": 0n,
    },
  ],
  "softBodyData": [],
  "soundScript": 0,
  "streakData": {},
  "windData": {},
}
`;

exports[`modl4 matches for ROOT: modl4-root 1`] = `
{
  "rootMotions": [],
}
`;
